plonky2x/src/frontend/ecc/ed25519/gadgets/verify.rs

```diff
@@ -0,0 +1,392 @@
+use num::BigUint;
+use plonky2::field::types::PrimeField;
+
+use crate::frontend::ecc::ed25519::curve::curve_types::{AffinePoint, Curve};
+use crate::frontend::ecc::ed25519::curve::ed25519::Ed25519;
+use crate::frontend::ecc::ed25519::field::ed25519_scalar::Ed25519Scalar;
+use crate::frontend::ecc::ed25519::gadgets::curve::{AffinePointTarget, CircuitBuilderCurve};
+use crate::frontend::ecc::ed25519::gadgets::eddsa::{
+    curta_batch_eddsa_verify_variable, EDDSASignatureTarget,
+};
+use crate::frontend::num::nonnative::nonnative::{CircuitBuilderNonNative, NonNativeTarget};
+use crate::frontend::vars::U32Variable;
+use crate::prelude::{BoolVariable, BytesVariable, CircuitBuilder, Field, PlonkParameters};
+
+pub struct DummySignatureTarget<C: Curve, const MAX_MESSAGE_LENGTH: usize> {
+    pub pubkey: AffinePointTarget<C>,
+    pub signature: EDDSASignatureTarget<C>,
+    pub message: BytesVariable<MAX_MESSAGE_LENGTH>,
+    pub message_byte_length: U32Variable,
+}
+
+// DUMMY_PRIVATE_KEY is [0u8; 32].
+pub const DUMMY_PUBLIC_KEY: [u8; 32] = [
+    59, 106, 39, 188, 206, 182, 164, 45, 98, 163, 168, 208, 42, 111, 13, 115, 101, 50, 21, 119, 29,
+    226, 67, 166, 58, 192, 72, 161, 139, 89, 218, 41,
+];
+pub const DUMMY_MSG: [u8; 32] = [0u8; 32];
+pub const DUMMY_MSG_LENGTH_BYTES: u32 = 32;
+pub const DUMMY_MSG_LENGTH_BITS: u32 = 256;
+
+// Produced by signing DUMMY_MSG with DUMMY_PRIVATE_KEY.
+pub const DUMMY_SIGNATURE: [u8; 64] = [
+    61, 161, 235, 223, 169, 110, 221, 24, 29, 190, 54, 89, 209, 192, 81, 196, 49, 240, 86, 165,
+    173, 106, 151, 166, 13, 92, 202, 16, 70, 4, 56, 120, 53, 70, 70, 30, 49, 40, 95, 197, 159, 145,
+    199, 7, 38, 66, 116, 80, 97, 226, 69, 29, 95, 243, 59, 204, 216, 195, 199, 77, 171, 202, 246,
+    10,
+];
+
+pub trait EDDSABatchVerify<L: PlonkParameters<D>, const D: usize> {
+    type Curve: Curve;
+    type ScalarField: PrimeField;
+
+    /// Dummy targets are used for inactive signatures and will always verify.
+    /// Invokers of batch_eddsa_verify must ensure the valid signatures are constrained correctly.
+    fn get_dummy_targets<const MAX_MESSAGE_BYTE_LENGTH: usize>(
+        &mut self,
+    ) -> DummySignatureTarget<Self::Curve, MAX_MESSAGE_BYTE_LENGTH>;
+
+    /// Verifies NUM_SIGS EdDSA signatures. Assumes all messages are the same length.
+    fn batch_eddsa_verify<const NUM_SIGS: usize, const MESSAGE_BYTE_LENGTH: usize>(
+        &mut self,
+        messages: Vec<BytesVariable<MESSAGE_BYTE_LENGTH>>,
```

Comment:
I think all the `Vec`s should be type [x; NUM_SIGS]

Reply:
sg, will use ArrayVariable for this